src/raftstore/store/store.rs

```diff
@@ -1456,6 +1617,257 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+                // skip early.
+                return Ok(());
+            }
+            let sibling_region = sibling_peer.region();
+
+            let min_index = peer.get_min_progress() + 1;
+            let low = cmp::max(min_index, state.get_min_index());
+            // TODO: move this into raft module.
+            let entries = if low >= state.get_commit() {
+                vec![]
+            } else {
+                self.region_peers[&region.get_id()]
+                    .get_store()
+                    .entries(low, state.get_commit(), NO_LIMIT)
+                    .unwrap()
+            };
+
+            let mut request = new_admin_request(sibling_region.get_id(), sibling_peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(sibling_region.get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::Merge);
+            admin.mut_merge().set_source(region.clone());
+            admin.mut_merge().set_commit(state.get_commit());
+            admin
+                .mut_merge()
+                .set_entries(RepeatedField::from_vec(entries));
+            request.set_admin_request(admin);
+            request
+        };
+        // Please note that, here assumes that the unit of network isolation is store rather than
+        // peer. So a quorum stores of souce region should also be the quorum stores of target
+        // region. Otherwise we need to enable proposal forwarding.
+        self.propose_raft_command(req, Callback::None);
+        Ok(())
+    }
+
+    fn rollback_pre_merge(&mut self, region: &metapb::Region) {
+        let req = {
+            let peer = &self.region_peers[&region.get_id()];
+            let state = peer.pending_merge.as_ref().unwrap();
+            let mut request = new_admin_request(region.get_id(), peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(peer.region().get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::RollbackPreMerge);
+            admin
+                .mut_rollback_pre_merge()
+                .set_commit(state.get_commit());
+            request.set_admin_request(admin);
+            request
+        };
+        self.propose_raft_command(req, Callback::None);
+    }
+
+    fn on_check_merge(&mut self, event_loop: &mut EventLoop<Self>) {
+        let merge_states = mem::replace(&mut self.merge_states, vec![]);
+        for region in &merge_states {
+            if let Err(e) = self.schedule_merge(region) {
+                info!(
+                    "[region {}] failed to schedule merge, rollback: {:?}",
+                    region.get_id(),
+                    e
+                );
+                self.rollback_pre_merge(region);
+            }
+        }
+        self.merge_states = merge_states;
```

Comment:
is there a possibility that other thread push new items into self.merge_states after mem::replace() and thus here will neglect these new items?

Reply:
As you can see from the method signature, it borrows `self` as `mut`, hence the method is accessed exclusively.